snappy/security.go

```diff
@@ -20,207 +20,835 @@
+	Syscalls []string `yaml:"syscalls,omitempty" json:"syscalls,omitempty"`
+}
+
+// SecurityAppArmorOverrideDefinition is used to override apparmor security
+// defaults
+type SecurityAppArmorOverrideDefinition struct {
+	ReadPaths    []string `yaml:"read-paths,omitempty" json:"read-paths,omitempty"`
+	WritePaths   []string `yaml:"write-paths,omitempty" json:"write-paths,omitempty"`
+	Abstractions []string `yaml:"abstractions,omitempty" json:"abstractions,omitempty"`
+}
+
+// SecurityOverrideDefinition is used to override apparmor or seccomp
+// security defaults
+type SecurityOverrideDefinition struct {
+	AppArmor *SecurityAppArmorOverrideDefinition `yaml:"apparmor" json:"apparmor"`
+	Seccomp  *SecuritySeccompOverrideDefinition  `yaml:"seccomp" json:"seccomp"`
+}
+
+// SecurityPolicyDefinition is used to provide hand-crafted policy
+type SecurityPolicyDefinition struct {
+	AppArmor string `yaml:"apparmor" json:"apparmor"`
+	Seccomp  string `yaml:"seccomp" json:"seccomp"`
+}
+
+// SecurityDefinitions contains the common apparmor/seccomp definitions
+type SecurityDefinitions struct {
+	// SecurityTemplate is a template like "default"
+	SecurityTemplate string `yaml:"security-template,omitempty" json:"security-template,omitempty"`
+	// SecurityOverride is a override for the high level security json
+	SecurityOverride *SecurityOverrideDefinition `yaml:"security-override,omitempty" json:"security-override,omitempty"`
+	// SecurityPolicy is a hand-crafted low-level policy
+	SecurityPolicy *SecurityPolicyDefinition `yaml:"security-policy,omitempty" json:"security-policy,omitempty"`
+
+	// SecurityCaps is are the apparmor/seccomp capabilities for an app
+	SecurityCaps []string `yaml:"caps,omitempty" json:"caps,omitempty"`
+}
+
+type securityAppID struct {
+	AppID   string
+	Pkgname string
+	Appname string
+	Version string
+}
+
+func aaPolicyDir() string {
+	// Templates and policy groups (caps)
+	aaPolicyDir := "/usr/share/apparmor/easyprof"
+	return filepath.Join(dirs.GlobalRootDir, aaPolicyDir)
+}
+
+func scPolicyDir() string {
+	scPolicyDir := "/usr/share/seccomp"
+	return filepath.Join(dirs.GlobalRootDir, scPolicyDir)
+}
+
+func aaFrameworkPolicyDir() string {
+	aaFrameworkPolicyDir := filepath.Join(policy.SecBase, "apparmor")
+	return filepath.Join(dirs.GlobalRootDir, aaFrameworkPolicyDir)
+}
+
+func scFrameworkPolicyDir() string {
+	scFrameworkPolicyDir := filepath.Join(policy.SecBase, "seccomp")
+	return filepath.Join(dirs.GlobalRootDir, scFrameworkPolicyDir)
+}
+
+func defaultPolicyVendor() string {
+	// FIXME: slightly ugly that we have to give a prefix here
+	return fmt.Sprintf("ubuntu-%s", release.Get().Flavor)
+}
+
+func defaultPolicyVersion() string {
+	return release.Get().Series
+}
+
+var dbusPathOkRegexp = regexp.MustCompile(`^[a-zA-Z0-9]$`)
+
+// Generate a string suitable for use in a DBus object
+func dbusPath(s string) string {
+	dbusStr := ""
+
+	for _, c := range strings.SplitAfter(s, "") {
```

Comment:
No, this is wrong. `strings.SplitAfter(s, "")` splits s into strings of length 1. Effectively it's like `[]rune(s)` except it's strings instead of individual runes.

What you need is a `[]byte`, because that's what you need to quote. You want simply `for _, c := range s`.